AX: [AX Thread Text APIs] Cache text run character widths up front to more accurately serve bounds APIs

hoffmanjoshua opened the pull request:
#### f7f3e11543724542dd187070c0cd5709ae8892d1
<pre>
AX: [AX Thread Text APIs] Cache text run character widths up front to more accurately serve bounds APIs
<a href="https://bugs.webkit.org/show_bug.cgi?id=290315">https://bugs.webkit.org/show_bug.cgi?id=290315</a>
<a href="https://rdar.apple.com/147752500">rdar://147752500</a>

Reviewed by Tyler Wilcock.

In the quest to improve the accuracy of the text marker bounds that are served off of the main thread,
this new approach caches the widths of each character in a run up-front. We do this a new method in
FontCascade to compute the width, and storing these widths in the text run. Then, when we serve the
bounds for a given range, we can sum up these advances (in the direction of the text orientation) to
build a more accurate bounds. This adds the extra benefit of more gracefully handling Horizontal and
Vertical text orientations.

Two new tests were added with this change. One test verifies bounds calculations when they span across
soft line breaks. The other tests bounds with mixed/upright vertical text.

* LayoutTests/accessibility-isolated-tree/TestExpectations:
* LayoutTests/accessibility/mac/bounds-for-range-expected.txt:
* LayoutTests/accessibility/mac/bounds-for-range-multiline-expected.txt: Added.
* LayoutTests/accessibility/mac/bounds-for-range-multiline.html: Added.
* LayoutTests/accessibility/mac/bounds-for-range-vertical-expected.txt: Added.
* LayoutTests/accessibility/mac/bounds-for-range-vertical.html: Added.
* LayoutTests/accessibility/mac/bounds-for-range.html:
* Source/WebCore/accessibility/AXTextMarker.cpp:
(WebCore::viewportRelativeFrameFromRuns):
* Source/WebCore/accessibility/AXTextRun.cpp:
(WebCore::AXTextRuns::localRect const):
* Source/WebCore/accessibility/AXTextRun.h:
(WebCore::AXTextRun::AXTextRun):
(WebCore::AXTextRun::advances const):
* Source/WebCore/accessibility/AccessibilityRenderObject.cpp:
(WebCore::AccessibilityRenderObject::textRuns):
* Source/WebCore/platform/graphics/FontCascade.cpp:
(WebCore::FontCascade::widthForCharacterInRun const):
* Source/WebCore/platform/graphics/FontCascade.h:

Canonical link: <a href="https://commits.webkit.org/292675@main">https://commits.webkit.org/292675@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/cfcc1aea7ec65d4e8391a7b2024bd45fb0c32b4c

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/96707 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/16321 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/6465 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/101780 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/47227 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/98752 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/16617 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/24760 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/73723 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/30923 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/99710 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/12508 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/87472 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/54038 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/12265 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/5274 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/46555 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/82368 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/5362 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/103803 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/23775 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/17340 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/82759 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/24025 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/83528 "Passed tests") | [❌ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/82158 "Found 3 new API test failures: /WebKitGTK/TestPrinting:/webkit/WebKitPrintOperation/print-errors, /WebKitGTK/TestPrinting:/webkit/WebKitPrintOperation/print, /WebKitGTK/TestPrinting:/webkit/WebKitPrintOperation/close-after-print (failure)") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/20631 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/26786 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/4325 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/17249 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/23737 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/28892 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/23396 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/26876 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/25137 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

twilco commented on LayoutTests/accessibility/bounds-for-range-multiline.html:
```diff
@@ -0,0 +1,38 @@
+    output += expectRectWithVariance("textChild.boundsForRange(0, 3)", -1, -1, 31, 23, 1); // "The"
+    output += expectRectWithVariance("textChild.boundsForRange(0, 19)", -1, -1, 168, 23, 1); // "The quick brown fox" (full width)
+    output += expectRectWithVariance("textChild.boundsForRange(16, 8)", -1, -1, 168, 46, 1); // "fox jumps" (full width)
+    output += expectRectWithVariance("textChild.boundsForRange(0, 42)", -1, -1, 168, 69, 1); // "The quick brown fox jumped over the lazy dog." (full width)
```
These comments are nice 🙏

twilco commented on Source/WebCore/accessibility/AXTextRun.cpp:
```diff
@@ -152,31 +152,30 @@ FloatRect AXTextRuns::localRect(unsigned start, unsigned end, float lineHeight,
-        CTFontOrientation ctFontOrientation = orientation == FontOrientation::Horizontal ? kCTFontOrientationHorizontal : kCTFontOrientationVertical;
-        return CTFontGetAdvancesForGlyphs(fontRef, ctFontOrientation, glyphs.begin(), advances.begin(), glyphs.size());
+    auto computeAdvance = [&] (const AXTextRun& run, unsigned offsetOfFirstCharacterInRun, unsigned startIndex, unsigned endIndex) {
+        RELEASE_ASSERT(startIndex <= endIndex);
```
Is this assert still the right thing to do? Seems like the subtraction we're protecting against is now `startIndex - offsetOfFirstCharacterInRun` and `endIndex - offsetOfFirstCharacterInRun`, so maybe asserting on those would be more relevant? What do you think?

twilco commented on Source/WebCore/accessibility/AXTextRun.cpp:
```diff
@@ -152,31 +152,30 @@ FloatRect AXTextRuns::localRect(unsigned start, unsigned end, float lineHeight,
-        return CTFontGetAdvancesForGlyphs(fontRef, ctFontOrientation, glyphs.begin(), advances.begin(), glyphs.size());
+    auto computeAdvance = [&] (const AXTextRun& run, unsigned offsetOfFirstCharacterInRun, unsigned startIndex, unsigned endIndex) {
+        RELEASE_ASSERT(startIndex <= endIndex);
+        auto characterAdvances = run.advances();
```
Using `auto` rather than `const auto&` creates a copy of the vector. Is a copy necessary here?

hoffmanjoshua replied:
Not necessary—updated to be a ref.

twilco commented on Source/WebCore/accessibility/AccessibilityRenderObject.cpp:
```diff
@@ -1405,12 +1407,15 @@ AXTextRuns AccessibilityRenderObject::textRuns()
     if (auto* renderLineBreak = dynamicDowncast<RenderLineBreak>(renderer.get())) {
         auto box = InlineIterator::boxFor(*renderLineBreak);
-        return { renderLineBreak->containingBlock(), { AXTextRun(box ? box->lineIndex() : 0, makeString('\n').isolatedCopy(), { lengthOneDomOffsets }) } };
+        // FIXME: this is probably wrong for the positions
```
Did you mean to include this comment? If so, could it be more descriptive? Which positions is this wrong for?

hoffmanjoshua replied:
Not intended to be included, removed this!

twilco commented on Source/WebCore/accessibility/AccessibilityRenderObject.cpp:
```diff
@@ -1405,12 +1407,15 @@ AXTextRuns AccessibilityRenderObject::textRuns()
         auto* containingBlock = renderer ? renderer->containingBlock() : nullptr;
-        return containingBlock ? AXTextRuns(containingBlock, { AXTextRun(0, String(span(objectReplacementCharacter)), { lengthOneDomOffsets }) }) : AXTextRuns();
+        uint16_t width = renderer ? static_cast<uint16_t>(frameRect().width()) : 0;
+        uint16_t height = renderer ? static_cast<uint16_t>(frameRect().height()) : 0;
```
`AccessibilityRenderObject::frameRect()` does a bit of work (e.g. some coordinate conversion). Maybe we could store it as a local to avoid doing that work twice?

twilco commented on Source/WebCore/accessibility/AccessibilityRenderObject.cpp:
```diff
@@ -1445,6 +1458,13 @@ AXTextRuns AccessibilityRenderObject::textRuns()
             else
                 lineString.append(character);
+
+            float characterWidth = (float)characterRects[i].width();
```
I think `static_cast<float>` is recommended over `(float)` wherever possible, as the former can catch more bugs at compile time.

twilco commented on Source/WebCore/accessibility/AccessibilityRenderObject.cpp:
```diff
@@ -1463,16 +1483,19 @@ AXTextRuns AccessibilityRenderObject::textRuns()
-            runs.append({ currentLineIndex, lineString.toString().isolatedCopy(), { std::exchange(textRunDomOffsets, { }) } });
+            runs.append({ currentLineIndex, lineString.toString().isolatedCopy(), { std::exchange(textRunDomOffsets, { }) }, WTFMove(characterWidths), lineHeight });
             lineString.clear();
+            characterWidths.clear();
```
Here we `WTFMove(characterWidths)`, and then call `.clear()` on it. I think that's a use-after-move bug, right? I think that we may want `std::exchange(characterWidths, { });` instead, which returns a `T&&` and replaces the value with an empty vector.

hoffmanjoshua replied:
Updated, thanks for calling this out!

minorninth commented on Source/WebCore/accessibility/AccessibilityRenderObject.cpp:
```diff
@@ -1445,6 +1458,13 @@ AXTextRuns AccessibilityRenderObject::textRuns()
                 lineString.append(character);
+
+            float characterWidth = (float)characterRects[i].width();
+            characterWidths.append(characterWidth + fmod(accumulatedDistanceFromStart, 1));
```
This is appending a float to characterWidths, implicitly casting to uint16_t. I'd prefer if this was an explicit floor or round, maybe even with a range check just in case?

hoffmanjoshua replied:
Fixed to be an explicit cast!

minorninth reviewed (commented):
looks great!

twilco commented on Source/WebCore/accessibility/AccessibilityRenderObject.cpp:
```diff
@@ -1463,16 +1483,19 @@ AXTextRuns AccessibilityRenderObject::textRuns()
 
     if (!lineString.isEmpty())
-        runs.append({ currentLineIndex, lineString.toString().isolatedCopy(), WTFMove(textRunDomOffsets) });
+        runs.append({ currentLineIndex, lineString.toString().isolatedCopy(), WTFMove(textRunDomOffsets), WTFMove(characterWidths), lineHeight });
```
Same comment about `std::exchange`.

hoffmanjoshua committed:
AX: [AX Thread Text APIs] Cache text run character widths up front to more accurately serve bounds APIs
https://bugs.webkit.org/show_bug.cgi?id=290315
rdar://147752500

Reviewed by Tyler Wilcock.

In the quest to improve the accuracy of the text marker bounds that are served off of the main thread,
this new approach caches the widths of each character in a run up-front. We do this a new method in
FontCascade to compute the width, and storing these widths in the text run. Then, when we serve the
bounds for a given range, we can sum up these advances (in the direction of the text orientation) to
build a more accurate bounds. This adds the extra benefit of more gracefully handling Horizontal and
Vertical text orientations.

Two new tests were added with this change. One test verifies bounds calculations when they span across
soft line breaks. The other tests bounds with mixed/upright vertical text.

* LayoutTests/accessibility-isolated-tree/TestExpectations:
* LayoutTests/accessibility/mac/bounds-for-range-expected.txt:
* LayoutTests/accessibility/mac/bounds-for-range-multiline-expected.txt: Added.
* LayoutTests/accessibility/mac/bounds-for-range-multiline.html: Added.
* LayoutTests/accessibility/mac/bounds-for-range-vertical-expected.txt: Added.
* LayoutTests/accessibility/mac/bounds-for-range-vertical.html: Added.
* LayoutTests/accessibility/mac/bounds-for-range.html:
* Source/WebCore/accessibility/AXTextMarker.cpp:
(WebCore::viewportRelativeFrameFromRuns):
* Source/WebCore/accessibility/AXTextRun.cpp:
(WebCore::AXTextRuns::localRect const):
* Source/WebCore/accessibility/AXTextRun.h:
(WebCore::AXTextRun::AXTextRun):
(WebCore::AXTextRun::advances const):
* Source/WebCore/accessibility/AccessibilityRenderObject.cpp:
(WebCore::AccessibilityRenderObject::textRuns):
* Source/WebCore/platform/graphics/FontCascade.cpp:
(WebCore::FontCascade::widthForCharacterInRun const):
* Source/WebCore/platform/graphics/FontCascade.h:

Canonical link: https://commits.webkit.org/292675@main
